enhancements/storage/csi-migration.md

```diff
@@ -0,0 +1,210 @@
+
+It is important to respect this ordering to avoid an undesired state of the components of volumes. For instance, when enabling the feature in the wrong order, volumes attached to nodes by the in-tree volume plugin cannot be detached by the CSI driver and will stay attached forever.
+
+In the same vein, when disabling the feature, volumes attached by the CSI driver cannot be detached by the in-tree volume plugin and will stay attached forever.
+
+In summary:
+
+* When the CSI migration is **enabled**, events should happen in this order:
+  1. Enable the feature gate in all control-plane components.
+  2. Once that's done, drain nodes one-by-one and start the kubelet with the feature gate enabled.
+
+* When the CSI migration is **disabled**, events should happen in this order:
+  1. One-by-one, drain the nodes and start the kubelet with the feature gate disabled.
+  2. Once that's done, disable the feature gate in all control-plane components.
+
+In order to achieve that, we propose two different approaches to be used at different times during the feature lifecycle.
+
+The first approach is intended to be used for is Tech Preview in OCP. The other approach should be used to once we graduate CSI migration to GA.
+
+### Tech Preview
+
+In OCP, *FeatureSets* can be used to aggregate one or more feature gates. Then, the resource [FeatureGate](https://github.com/openshift/api/blob/dca637550e8c80dc2fa5ff6653b43a3b5c6c810c/config/v1/types_feature.go#L9-L21) can be used to enable and disable *FeatureSets*.
+
+With that in mind, we propose to:
+
+1. Create two [new FeatureSets](https://github.com/openshift/api/blob/master/config/v1/types_feature.go#L25-L43) to support CSI migration: `CSIMigrationNode` and `CSIMigrationControlPlane`.
+   * Both *FeatureSets* will enable the **same** feature gates:
+     - `CSIMigration`
+     - `CSIMIgrationAWS`
+     - `CSIMigrationGCE`
+     - `CSIMigrationAzureDisk`
+     - `CSIMigrationAzureFile`
+     - `CSIMigrationvSphere`
+     - `CSIMigrationOpenStack`
+   * The machine-config-operator (MCO) will **ignore** the `CSIMigrationControlPlane` **FeatureSet**.
+   * On the other hand, all operators will react to the `CSIMigrationNode` *FeatureSet*, including control-plane operators.
+2. To enable CSI Migration for any in-tree plugin, the cluster administrator should:
+   * First, enable the CSI migration feature flags in all control-plane components:
+     - Add the `CSIMigrationControlPlane` *FeatureSet* to the `featuregates/cluster` object:
+       ```shell
+       $ oc edit featuregates/cluster
+       (...)
+       spec:
+         featureSet: CSIMigrationControlPlane
+       ```
+     - With the exception of MCO, all operators will recognize this *FeatureSet* and will initialize their operands with the associated feature gates.
+   * Second, once all control-plane components have restarted, enable the CSI migration feature flags in the kubelet:
```

Comment:
How do you ensure that the order isn't wrong and that the prerequisites are fulfilled? As mentioned before several times the ordering is **critical** piece here, giving that control to a cluster administrator is dilluting the responsibility. I'd personally prefer that inter-dependencies are tracked by MCO monitoring other operators conditions, similarly how this was laid out in https://github.com/openshift/enhancements/pull/463. For that purpose all the control-plane operators could set `CSIMigrationControlPlan` condition when the rollout succeeds, and MCO on ensuring that all pre-reqs have been met will proceed with applying the necessary changes to kubelet. This will ensure stronger guarantees than those presented here.

Reply:
We've dropped this approach and re-worked this enhancement. PTAL